modules/@apostrophecms/permission/index.js

```diff
@@ -67,28 +103,82 @@ module.exports = {
 
       // Returns a MongoDB criteria object that retrieves only documents
       // the user is allowed to perform `action` on.
-      //
-      // If a permission is not specific to particular documents,
-      // for instance `view-draft`, `docOrType` may be omitted.
 
       criteria(req, action) {
-        if (req.user) {
-          // For now, users can do anything
+        const role = req.user && req.user.role;
+        if (role === 'admin') {
           return {};
         }
-        if (action !== 'view') {
-          // Public can only view for now
+        const restrictedTypes = Object.keys(self.apos.doc.managers).filter(name => self.apos.doc.getManager(name).isAdminOnly());
+        if (action === 'view') {
+          if (role === 'guest') {
+            return {
+              aposMode: {
+                $in: [ null, 'published' ]
+              },
+              visibility: {
+                $in: [ 'public', 'loginRequired' ]
+              },
+              type: {
+                $nin: restrictedTypes
+              }
+            };
+          } else if ((role === 'contributor') || (role === 'editor')) {
+            return {
+              type: {
+                $nin: restrictedTypes
+              }
+            };
+          } else {
+            return {
+              aposMode: {
+                $in: [ null, 'published' ]
+              },
+              visibility: 'public',
+              type: {
+                $nin: restrictedTypes
+              }
+            };
+          }
+        } else if (action === 'edit') {
+          if (role === 'contributor') {
+            return {
+              aposMode: {
+                $in: [ null, 'draft' ]
+              },
+              type: {
+                $nin: restrictedTypes
+              }
+            };
+          } else if (role === 'editor') {
+            return {
+              type: {
+                $nin: restrictedTypes
+              }
+            };
+          } else {
+            return {
+              _id: 'thisIdWillNeverMatch'
```

Comment:
Can't we throw an error? Somehow they've tried to initiate an editing action without being a contrib, editor, or admin. Something is actually wrong there.

Reply:
I wanted to make this a little less brittle than that, in case someone is using the super pattern on it; exception throwing is a little expensive if it has to happen every time and then get caught and turned into something else. Whereas changing the return value would be lightweight and speedy.